### PR TITLE
kool create command

### PR DESCRIPTION
**Pull request by leandroncbrito**
https://github.com/kool-dev/kool/issues/17

New command "create"

**Parameters:** 
1. preset/framework name
2. project/folder name

Ex: kool create laravel my-app

**New issue improvement:** 
Instead of entering the preset name, we might show the preset menu and let the user choose.

### CODE REVIEW

**Comment by dbpolito** [comment]
@leandroncbrito after you create, it's already having the kool files there? or you still need to `init/preset` afterwards?

I think we can safely create and init.

File: cmd/create.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// KoolCreate holds handlers and functions to implement the preset command logic
+type KoolCreate struct {
+	DefaultKoolService
+	parser     presets.Parser
+	koolDocker builder.Command
```
Comment (danielsuguimoto) [inline cmd/create.go:16]: Instead of a command builder calling `kool docker`, you can use the `KoolDocker` service.

File: cmd/create.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		&presets.DefaultParser{Presets: presets.GetAll()},
+		builder.NewCommand("kool", "docker"),
+	}
+}
+
+// Execute runs the exec logic with incoming arguments.
+func (c *KoolCreate) Execute(args []string) (err error) {
+	if !c.IsTerminal() {
+		c.koolDocker.AppendArgs("-T")
+	}
```
Comment (danielsuguimoto) [inline cmd/create.go:41]: `KoolDocker` already handles it if you use it instead of a command builder.

File: cmd/presets/fake_parser.go [hunk 3]
```diff
@@ -19,6 +20,13 @@ func (f *FakeParser) Exists(preset string) (exists bool) {
 	return
 }
 
+// GetCreateCommand gets the command to create a new project
+func (f *FakeParser) GetCreateCommand(preset string) (cmd string, err error) {
```
Comment (danielsuguimoto) [inline cmd/presets/fake_parser.go:24]: Add a test case for this new function in `fake_parser_test.go`

**Review by danielsuguimoto — CHANGES_REQUESTED** [review]

**Comment by danielsuguimoto** [comment]
> @leandroncbrito after you create, it's already having the kool files there? or you still need to `init/preset` afterwards?
> 
> I think we can safely create and init.

I like the idea too. The only question I have: should we add an option for creating the project only, without copying the preset files?

**Comment by dbpolito** [comment]
Yeah but i think it makes sense creating it as default... maybe a `--no-preset` flag to disable it?

**Comment by fabriciojs** [comment]
> Yeah but i think it makes sense creating it as default... maybe a `--no-preset` flag to disable it?

No need for this flag IMO. Design wise, we want people to use our environment. Even if they do not want to, having the files installed is not a big deal - no issues should arise from that.

**Comment by leandroncbrito** [comment]
Yeah, I like the idea. 
I'll make the code improvements that Jun left and take a look at this change.

**Review by danielsuguimoto — COMMENTED** [review]
Nice! I would add just one test case to the `create` command. If we add a test case and mock the preset fake parser to return true on the method `Exists` (set `MockExists` to true), then we can improve the test coverage. This way I guess all checks for this PR will pass.

**Review by danielsuguimoto — APPROVED** [review]
LGTM! Thx!

**Review by danielsuguimoto — CHANGES_REQUESTED** [review]
@leandroncbrito sorry, I was almost forgetting about the docs. Please, run `kool run make-docs` to generate the docs markdown to this new command, then I think it's ready to go.

**Comment by leandroncbrito** [comment]
> @leandroncbrito sorry, I was almost forgetting about the docs. Please, run `kool run make-docs` to generate the docs markdown to this new command, then I think it's ready to go.

done

File: docs/4-Commands/0-kool.md [hunk 4]
```diff
@@ -16,10 +16,11 @@ Complete documentation is available at https://kool.dev/docs
 ```
 
 ### SEE ALSO
+
+* [kool create](kool-create.md)	 - Create a new project using preset
 * [kool docker](kool-docker.md)	 - Creates a new container and runs the command in it.
 * [kool exec](kool-exec.md)	 - Execute a command within a running service container
 * [kool info](kool-info.md)	 - Prints out information about kool setup (like environment variables)
-* [kool init](kool-init.md)	 - [DEPRECATED] Proxies preset command
```
Comment (danielsuguimoto) [inline docs/4-Commands/0-kool.md:22]: I guess the script to generate the docs has an issue with this proxied command and it's removing it by mistake. It will be removed only at version `2.0.0`, so we should keep it here, for now...

**Review by danielsuguimoto — CHANGES_REQUESTED** [review]

File: docs/4-Commands/kool-init.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ## kool init
 
-[DEPRECATED] Initialize kool preset in the current working directory. If no preset argument is specified you will be prompted to pick among the existing options.
```
Comment (leandroncbrito) [inline docs/4-Commands/kool-init.md:3]: @danielsuguimoto is it the same case here? Should I revert this change?

Comment (danielsuguimoto) [reply docs/4-Commands/kool-init.md:3]: @leandroncbrito yep

**Review by danielsuguimoto — APPROVED** [review]

**Comment by danielsuguimoto** [comment]
LGTM! What do you think @dbpolito @fabriciojs ?

**Review by dbpolito — APPROVED** [review]

File: cmd/presets/parser.go [hunk 6]
```diff
@@ -29,6 +30,21 @@ func (p *DefaultParser) Exists(preset string) (exists bool) {
...
+// ErrCreateCommandtNotFoundOrEmpty error throwed when did not find the preset create command or it's empty
+var ErrCreateCommandtNotFoundOrEmpty = errors.New("create command not found or empty")
+
+// GetCreateCommand gets the command to create a new project
+func (p *DefaultParser) GetCreateCommand(preset string) (cmd string, err error) {
+	cmd = p.Presets[preset]["preset_create"]
+
+	if cmd == "" {
+		err = ErrCreateCommandtNotFoundOrEmpty
+		return
```
Comment (fabriciojs) [inline cmd/presets/parser.go:42]: no need to have this `return`

File: cmd/create.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+	)
+
+	rootCmd.AddCommand(createCmd)
+}
+
+// NewKoolCreate creates a new handler for create logic
+func NewKoolCreate() *KoolCreate {
+	return &KoolCreate{
+		*newDefaultKoolService(),
+		&presets.DefaultParser{Presets: presets.GetAll()},
```
Comment (fabriciojs) [inline cmd/create.go:33]: I guess would be good if we could do this in a lazy fashion... Right now for absolutely every single execution of `kool` it will have to call NewKoolCreate on the `init()` of this file, and allocate the memory for the presets maps - which basically undo a change we made a while ago - https://github.com/kool-dev/kool/commit/788fc6b993b3cc8e173b961f60868911ac22c948

Not sure how to make it lazy as it is now though, need to think 🤔

File: cmd/create.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+	return
+}
+
+// NewCreateCommand initializes new kool create command
+func NewCreateCommand(create *KoolCreate) (createCmd *cobra.Command) {
+	createCmd = &cobra.Command{
+		Use:   "create [preset] [project]",
+		Short: "Create a new project using preset",
+		Args:  cobra.MinimumNArgs(2),
```
Comment (fabriciojs) [inline cmd/create.go:75]: we only expect the `2` arguments right? I guess we can make this `cobra.ExactNArgs` (not sure the validation rule name)

File: cmd/create.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+		*NewKoolPreset(),
+	}
+}
+
+// Execute runs the create logic with incoming arguments.
+func (c *KoolCreate) Execute(originalArgs []string) (err error) {
+	preset := originalArgs[0]
+	dir := originalArgs[1]
+
+	if !c.parser.Exists(preset) {
```
Comment (fabriciojs) [inline cmd/create.go:44]: maybe we only instantiate the `c.parser` here, which would make it lazy enough already.

File: cmd/presets/parser.go [hunk 10]
```diff
@@ -29,6 +30,21 @@ func (p *DefaultParser) Exists(preset string) (exists bool) {
 	return
 }
 
+// ErrCreateCommandtNotFoundOrEmpty error throwed when did not find the preset create command or it's empty
+var ErrCreateCommandtNotFoundOrEmpty = errors.New("create command not found or empty")
+
+// GetCreateCommand gets the command to create a new project
+func (p *DefaultParser) GetCreateCommand(preset string) (cmd string, err error) {
```
Comment (fabriciojs) [inline cmd/presets/parser.go:37]: just a thought for a later iteration - there might be some preset which requires more than a single command to "create" a new project... we probably could make this abstract, like we do for kool.yml `scripts:  task: single | multiline` do the same for the `preset_create`.

Comment (fabriciojs) [reply cmd/presets/parser.go:37]: And then here for example we could already return a Builder or something, to already encapsulate the runnable create commands.

File: cmd/presets/presets.go [hunk 11]
```diff
@@ -193,6 +195,7 @@ networks:
 		".dockerignore": `/node_modules
 `,
 		"preset_language": "javascript",
+		"preset_create": "kooldev/node:14-nest nest new",
```
Comment (fabriciojs) [inline cmd/presets/presets.go:198]: Having the `preset_create` standard like this, making it mandatory to run only `kool docker` commands, I guess is an unnecessary constraint... why can't we simply leave it for the user the whole command? For example all these could simple be prefixed with `kool docker <image> <commands...>`, giving it more flexibility IMO.

**Review by fabriciojs — CHANGES_REQUESTED** [review]
@leandroncbrito looking great! thanks for all the work on this!

I left a handful of comments, some we might leave for a feature iteration, others we may discuss further. Let me know what you think!

File: cmd/create.go (hunk 7 above)
Comment (danielsuguimoto) [reply cmd/create.go:33]: @fabriciojs @leandroncbrito instead of adding the presets on `NewKoolCreate`, we can add the `LoadPresets` function and call it inside create command execution. The same can be done on preset/init

**Review by fabriciojs — APPROVED** [review]
